File: May-LeetCoding-Challenge/09-Valid-Perfect-Square/Valid-Perfect-Square.java
class Solution {
    // Binary search solution, make the mid * mid closer to num in every loop
    public boolean isPerfectSquare(int num) {
        long left = 1, right = num;
        
        while(left <= right) {
            long mid = (left + right) / 2;
            
            if(mid * mid == num) return true;
            
            if(mid * mid < num ) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return false;
    }
}

File: May-LeetCoding-Challenge/09-Valid-Perfect-Square/Valid-Perfect-Square.cs
public class Solution {
    public bool IsPerfectSquare(int num) {
        int temp=num;
        int x=num/2;
        if (x<2)
        {
            return true;
        }
        {
        while ( x > num/x )
        {
            x=temp/2;
            temp=x;
        }
        for(int i = x; i<2*x ; i++)
        {
            if (i*i == num)
            {
                return true;
            }
        
        }
        return false;   
            }
    }
}
